src/Build/Construction/Solution/ProjectInSolution.cs

```diff
@@ -342,7 +342,7 @@ internal bool CanBeMSBuildProjectFile(out string errorMessage)
                     // not an MSBuild file. It will however have ToolsVersion="2.0" which is not supported with an empty
                     // schema. This is not a great solution, but it should cover the customer reported issue. See:
                     // https://github.com/dotnet/msbuild/issues/2064
-                    if (emptyNamespace && !projectElementInvalid && mainProjectElement.GetAttribute("ToolsVersion") != "2.0")
+                    if (emptyNamespace && !projectElementInvalid && mainProjectElement.GetAttribute("ToolsVersion") != "2.0" && !Path.GetExtension(AbsolutePath).Equals(".dwproj", StringComparison.OrdinalIgnoreCase))
```

Comment:
Are there any more robust checks we could apply, like "defines a namespace for `http://schemas.microsoft.com/DataWarehouse/Designer/1.0`"?

Reply:
Basically I'd like to future-proof against "we have a new version of dwproj tooling and dwproj is an msbuild project now" which has happened with other project types.

Reply:
I'm not sure that's really more future-proof. If the format for dwproj changes, and it no longer defines that namespace (or doesn't do whatever else we would do to check if it's msbuild-y), #2064 would come back unexpectedly. Of note, it would come back for a lot of people all at once. If dwproj switches to being msbuild-y, we would be wrong with the check here, but that's more of an expected issue, which should make it easier to debug. If it becomes msbuild-y without changing the thing we look at, we'd have the same issue as with this extension check.

An alternative could be to check for both dwproj _and_ make sure there are at least some unexpected child elements under the project element. That would probably be the most robust, but it would also lead to confusing cases with MSBuild mostly working for dwproj but not always giving proper error messages. This version seems easiest for us to debug and fix if dwproj becomes msbuild-y.

Reply:
I do not understand your concern here:

> If the format for dwproj changes, and it no longer defines that namespace (or doesn't do whatever else we would do to check if it's msbuild-y), #2064 would come back unexpectedly. Of note, it would come back for a lot of people all at once.

Yes, there are future changes that could break our detection. But how is that different from today?

> If dwproj switches to being msbuild-y, we would be wrong with the check here, but that's more of an expected issue, which should make it easier to debug.

It would mean though that the owners of dwproj couldn't make forward progress without getting an MSBuild change, and at that time wouldn't the change need to be what I'm asking for here, since there would be the possibility of encountering old and new dwproj files?

> An alternative could be to check for both dwproj _and_ make sure there are at least some unexpected child elements under the project element. That would probably be the most robust, but it would also lead to confusing cases with MSBuild mostly working for dwproj but not always giving proper error messages.

I don't understand this concern, can you please elaborate?

I would be fine scoping the check to only the extension dwproj.

Reply:
The two proposals here are:
1. Explicitly check for a dwproj extension
2. Check for something that indicates that this is a dwproj without checking the extension explicitly.

The two problems we want to avoid are:
1. We think something is not an MSBuild project, but it is.
2. We think something is an MSBuild project, but it is not.

If there is a future change to cause one of those problems, I want it to be as obvious as possible. "I upgrade my .dwproj to a new version that can now be built by MSBuild, whereas the previous version could not be" is a place where it's obvious the user should verify that MSBuild can handle that, whereas "I updated from schemas.microsoft.com/.../1.0 to 1.1" (with dwproj still not MSBuild-y) makes it seem like the problem is probably with the new schema, which would lead users down the wrong path.

It is true that with an explicit extension check, there is nothing the user can do to indicate to MSBuild that their dwproj is buildable. On the other hand, if we check for something optional, the user can decide not to set it, and we would be wrong again without dwproj changing anything. If we check for something required, the user can't remove it anyway, so it doesn't help.

If dwproj switches to being MSBuild-y, you are correct that we would have to have something other than the extension we can look at. However, at that point, we would have to talk with the owners to see what they are willing to commit to. Arbitrarily choosing something makes us liable to being broken. Engaging them in a discussion about it now constrains them with no added value unless they decide to make dwproj MSBuild-y. They haven't done that for years, and I see no reason to assume that's about to change.